x/ibc/06-solomachine/types/client_state.go

```diff
@@ -0,0 +1,438 @@
+	return Initialize(msg.GetClientID(), msg.ConsensusState)
+}
+
+// Initialize creates an unfrozen client with the initial consensus state
+func Initialize(id string, consensusState ConsensusState) (ClientState, error) {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}, nil
+}
+
+// NewClientState creates a new ClientState instance.
+func NewClientState(id string, consensusState ConsensusState) ClientState {
+	return ClientState{
+		ID:             id,
+		Frozen:         false,
+		ConsensusState: consensusState,
+	}
+}
+
+// GetID returns the solo machine client state identifier.
+func (cs ClientState) GetID() string {
+	return cs.ID
+}
+
+// GetChainID returns an empty string.
+func (cs ClientState) GetChainID() string {
+	return ""
+}
+
+// ClientType is Solo Machine.
+func (cs ClientState) ClientType() clientexported.ClientType {
+	return clientexported.SoloMachine
+}
+
+// GetLatestHeight returns the latest sequence number.
+func (cs ClientState) GetLatestHeight() uint64 {
+	return cs.ConsensusState.Sequence
+}
+
+// IsFrozen returns true if the client is frozen
+func (cs ClientState) IsFrozen() bool {
+	return cs.Frozen
+}
+
+// Validate performs basic validation of the client state fields.
+func (cs ClientState) Validate() error {
+	if err := host.ClientIdentifierValidator(cs.ID); err != nil {
+		return err
+	}
+	return cs.ConsensusState.ValidateBasic()
+}
+
+// VerifyClientConsensusState verifies a proof of the consensus state of the
+// Solo Machine client stored on the target machine.
+func (cs ClientState) VerifyClientConsensusState(
+	store sdk.KVStore,
+	cdc *codec.Codec,
+	root commitmentexported.Root,
+	_ uint64,
+	counterpartyClientIdentifier string,
+	consensusHeight uint64,
+	prefix commitmentexported.Prefix,
+	proof commitmentexported.Proof,
+	consensusState clientexported.ConsensusState,
+) error {
+	clientPrefixedPath := "clients/" + counterpartyClientIdentifier + "/" + host.ConsensusStatePath(consensusHeight)
+	path, err := commitmenttypes.ApplyPrefix(prefix, clientPrefixedPath)
+	if err != nil {
+		return err
+	}
+
+	if cs.IsFrozen() {
+		return clienttypes.ErrClientFrozen
+	}
+
+	// cast the proof to a signature proof
+	signatureProof, ok := proof.(commitmenttypes.SignatureProof)
+	if !ok {
+		return sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "proof type %T is not type SignatureProof", proof)
+	}
+
+	bz, err := cdc.MarshalBinaryBare(consensusState)
+	if err != nil {
+		return err
+	}
+
+	// value = sequence + path + consensus state
+	value := append(
+		combineSequenceAndPath(cs.ConsensusState.Sequence, path),
+		bz...,
+	)
+	if !cs.ConsensusState.PubKey.VerifyBytes(value, signatureProof.Signature) {
+		return sdkerrors.Wrap(clienttypes.ErrFailedClientConsensusStateVerification, "failed to verify proof against current public key, sequence, and consensus state")
+	}
+
+	cs.ConsensusState.Sequence++
+	setClientState(store, cs)
```

Comment:
@cwgoes, it's weird to me that the consensus state is stored inside the client state, yet we have another SetConsensusState function for tendermint clients. Why the difference with ICS07?

Reply:
Think the `ConsensusState` should be removed from here and replace with `LatestSequence`. Then we should be setting consensus state as 02-client expects

Reply:
The nomenclature is a bit confusing, the only "consensus state" a solo machine has is a public key and a nonce, there is no need to set it each height since the state isn't Merklized.

@AdityaSripal's suggestion seems fine to me, probably a bit clearer.